x-pack/elastic-agent/pkg/agent/cmd/run.go

```diff
@@ -47,6 +48,12 @@ func run(flags *globalFlags, streams *cli.IOStreams) error {
 		return err
 	}
 
+	locker := application.NewAppLocker(paths.Data())
+	if err := locker.TryLock(); err != nil {
+		return err
+	}
+	defer locker.Unlock()
```

Comment:
can you make sure this is handled even if we are killed. defer statements are skipped if SIGINT or SIGTERM are received and it can prevent us from restarting

Reply:
I have verified that this does get called in all the cases defined below with signals.

```
signals := make(chan os.Signal, 1)
signal.Notify(signals, syscall.SIGINT, syscall.SIGKILL, syscall.SIGTERM, syscall.SIGQUIT)

<-signals
```

So the defer does get called. I did find a bug in periodic that was preventing `app.Start` from returning to catch the signals. I have fixed that in my most recent commit.